# Script to run analyses for G6PD age of diagnosis project using G&H data prepared by G&H team
# June 2024

library(arrow)
library(ggplot2)

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

setwd("#####")

##### DATA PREPARATION #####

# Read in G&H genotype and phenotype data file generated by G&H team
alldata <- read_feather("/#####/X.154534419_phenotype.feather")
# Keep only IDs with non-missing genotype for variant and sex
alldata <- alldata[!is.na(alldata$med_variant) & !is.na(alldata$sex_corrected),]
alldata <- alldata[!duplicated(alldata$NHS_ID),]
length(unique(alldata$NHS_ID))

# Recode genotype variable to represent minor allele count
alldata$genotype <- abs(alldata$med_variant - 2)

# Define sex as a factor variable with labels
alldata$Sex <- factor(alldata$sex_corrected, levels = 1:2, labels = c("Male", "Female"))
# Check if any males with genotype of 1 - possible Klinefelters syndrome
length(unique(alldata[alldata$Sex == "Male" & alldata$genotype == 1,]$NHS_ID))

# Read in self-reported ethnicity from questionnaire data
selfdata <- read.csv("/#####/2024_04_30__S1QST_redacted.tab", header = TRUE, sep = "\t")
selfdata$S1QST_Oragene_ID <- format(selfdata$S1QST_Oragene_ID, scientific = FALSE)
selfdata$S1QST_Oragene_ID <- paste0("GNH-", selfdata$S1QST_Oragene_ID)
selfdata$ethnicity <- NA
selfdata[!is.na(selfdata$S1QST_EthnicGroup) & selfdata$S1QST_EthnicGroup == 1,]$ethnicity <- "Bangladeshi"
selfdata[!is.na(selfdata$S1QST_EthnicGroup) & selfdata$S1QST_EthnicGroup == 2,]$ethnicity <- "Pakistani"
selfdata[(!is.na(selfdata$S1QST_EthnicGroup) & selfdata$S1QST_EthnicGroup == 3),]$ethnicity <- "Other"
selfdata[is.na(selfdata$S1QST_EthnicGroup),]$ethnicity <- "Missing"
selfdata <- selfdata[c("S1QST_Oragene_ID", "ethnicity")]
selfdata <- selfdata[!duplicated(selfdata$S1QST_Oragene_ID),]
alldata <- merge(alldata, selfdata, by.x = "orageneID", by.y = "S1QST_Oragene_ID")

## Generate summary characteristics of Genes & Health cohort

length(unique(alldata$NHS_ID))

nrow(alldata[alldata$Sex == "Male",])
nrow(alldata[alldata$Sex == "Female",])

nrow(alldata[alldata$ethnicity == "Bangladeshi",])
nrow(alldata[alldata$ethnicity == "Pakistani",])

mean(alldata$HbA1c_age, na.rm = TRUE)
sd(alldata$HbA1c_age, na.rm = TRUE)

mean(alldata$bmi, na.rm = TRUE)
sd(alldata$bmi, na.rm = TRUE)

mean(alldata$HbA1c, na.rm = TRUE)
sd(alldata$HbA1c, na.rm = TRUE)

mean(alldata$random_glucose, na.rm = TRUE)
sd(alldata$random_glucose, na.rm = TRUE)

nrow(alldata[alldata$G6PD == 1,])

## Calculate MAF for the IDs in the cohort - for self-reported ethnicity

mafdata <- data.frame(matrix(NA, nrow = 2, ncol = 2))
colnames(mafdata) <- c("Sex", "MAF")
mafdata$Sex <- c("Male", "Female")
for (i in 1:nrow(mafdata)) {
  subdata <- alldata[alldata$Sex == mafdata$Sex[i],]
  mafdata$MAF[i] <- sum(subdata$genotype)/(2*nrow(subdata))
}
mafdata$Prevalence <- 100*mafdata$MAF

write.table(mafdata, "prevalence_G6PD_variants_GnH.txt", col.names = TRUE, row.names = FALSE, sep = "\t", quote = FALSE)

## Determine the prevalence of G6PD deficiency diagnoses in variant carriers

myresults <- data.frame(matrix(NA, nrow = 5, ncol = 4))
colnames(myresults) <- c("Sex", "Genotype", "N", "G6PD_deficiency")
myresults$Sex <- c(rep("Male", 2), rep("Female", 3))
myresults$Genotype <- c(0,2,0:2)
for (i in 1:nrow(myresults)) {
  subdata <- alldata[alldata$Sex == myresults$Sex[i],]
  subdata <- subdata[subdata$genotype == myresults$Genotype[i],]
  myresults$N[i] <- length(unique(subdata$NHS_ID))
  myresults$G6PD_deficiency[i] <- length(unique(subdata[subdata$G6PD == 1,]$NHS_ID))
}

write.table(myresults, file = "genotype_g6pd_GnH.txt", col.names = TRUE, row.names = FALSE, sep= "\t", quote = FALSE)

### Compare HbA1c and random glucose across genotypes ###

# Recode the male homozygote genotype from 2 to 1 for ease of the regression models - to keep as continuous variable (makes no difference)
alldata[alldata$Sex == "Male" & alldata$genotype == 2,]$genotype <- 1

## Generate summary characteristics of cohort

# Find number whose only measurement is taken after diabetes (any) diagnosis or during gestation or postpartum period of pregnancy
length(unique(alldata[alldata$all_diabetes == 1,]$NHS_ID))

## Apply exclusion criteria for HbA1c/glucose analyses

# Exclude those with missing HbA1c measurements
nodiabdata <- alldata[!is.na(alldata$HbA1c),]

# Exclude those whose only measurement is taken after diabetes (any) diagnosis or during gestation or postpartum period of pregnancy
nodiabdata <- nodiabdata[nodiabdata$all_diabetes == 0,]

## Plot HbA1c and RG violin plots for genotypes for each sex

# Generate genotype and sample size labels required for plots
nodiabdata$label <- nodiabdata$sub_label <- nodiabdata$label2 <- nodiabdata$sub_label2 <- nodiabdata$N <- nodiabdata$sub_N <- NA
nodiabdata$Sex <- factor(nodiabdata$Sex, levels = c("Male", "Female"))
hba1cdata <- rgdata <- nodiabdata
hba1cdata <- nodiabdata[!is.na(nodiabdata$HbA1c),]
rgdata <- nodiabdata[!is.na(nodiabdata$random_glucose),]

tabdata <- data.frame(matrix(NA, nrow = 5, ncol = 3))
colnames(tabdata) <- c("Sex", "Genotype", "Status")
tabdata$Sex <- c(rep("Male", 2), rep("Female", 3))
tabdata$Genotype <- c(0, 1, 0:2)
tabdata$Status <- c("No mutations", "Hemizygote", "No mutations", "Heterozygote", "Homozygote")

for (i in 1:nrow(tabdata)) {
  sex <- tabdata$Sex[i]
  genotype <- tabdata$Genotype[i]
  status <- tabdata$Status[i]
  subdata <- nodiabdata[nodiabdata$Sex == sex & nodiabdata$genotype == genotype,]
  HbA1c.N <- length(unique(subdata[!is.na(subdata$HbA1c),]$NHS_ID))
  HbA1c.N <- prettyNum(HbA1c.N, big.mark = ",", scientific = FALSE)
  RG.N <- length(unique(subdata[!is.na(subdata$random_glucose),]$NHS_ID))
  RG.N <- prettyNum(RG.N, big.mark = ",", scientific = FALSE)
  if (HbA1c.N != 0) {
    hba1cdata[hba1cdata$Sex == sex & hba1cdata$genotype == genotype,]$label <- paste0(status, "\n N = ", HbA1c.N)
    hba1cdata[hba1cdata$Sex == sex & hba1cdata$genotype == genotype,]$label2 <- paste0(status, " (N = ", HbA1c.N, ")")
    hba1cdata[hba1cdata$Sex == sex & hba1cdata$genotype == genotype,]$N <- length(unique(subdata[!is.na(subdata$HbA1c),]$NHS_ID))
  }
  if (RG.N != 0) {
    rgdata[rgdata$Sex == sex & rgdata$genotype == genotype,]$label <- paste0(status, "\n N = ", RG.N)
    rgdata[rgdata$Sex == sex & rgdata$genotype == genotype,]$label2 <- paste0(status, " (N = ", RG.N, ")")
    rgdata[rgdata$Sex == sex & rgdata$genotype == genotype,]$N <- length(unique(subdata[!is.na(subdata$random_glucose),]$NHS_ID))
  }
}

hba1cdata <- hba1cdata[order(hba1cdata$Sex, hba1cdata$genotype),]
hba1cdata$label <- factor(hba1cdata$label, levels = unique(hba1cdata$label))
hba1cdata$label2 <- factor(hba1cdata$label2, levels = unique(hba1cdata$label2))
hba1cdata$sub_label <- factor(hba1cdata$sub_label, levels = unique(hba1cdata$sub_label))
hba1cdata$sub_label2 <- factor(hba1cdata$sub_label2, levels = unique(hba1cdata$sub_label2))

rgdata <- rgdata[order(rgdata$Sex, rgdata$genotype),]
rgdata$label <- factor(rgdata$label, levels = unique(rgdata$label))
rgdata$label2 <- factor(rgdata$label2, levels = unique(rgdata$label2))
rgdata$sub_label <- factor(rgdata$sub_label, levels = unique(rgdata$sub_label))
rgdata$sub_label2 <- factor(rgdata$sub_label2, levels = unique(rgdata$sub_label2))

# Create function to generate data and labels required and produce violin plots
violinplot <- function(measure) {
  if (!(measure %in% c("HbA1c", "random_glucose"))) {
    stop("Missing or incorrect function input provided")
  }
  # Subset the appropriate dataframe
  if (measure == "HbA1c") {
    subdata <- hba1cdata
  } else {
    subdata <- rgdata
  }
  subdata$measure <- unlist(subdata[colnames(subdata) == measure])
  subdata <- subdata[!is.na(subdata$measure),]
  # Calculate the truncation limits based on MAD
  highlimit <- median(subdata$measure) + (5*mad(subdata$measure))
  lowlimit <- max(c(min(subdata$measure), (median(subdata$measure) - (5*mad(subdata$measure)))))
  if (nrow(subdata[subdata$measure > highlimit,]) > 0) {
    nabove <- aggregate(measure ~ label, data = subdata[subdata$measure > highlimit,], length)
    maxvals <- aggregate(measure ~ label, data = subdata, max)
    # Create dataframe containing labels and variables defining location of labels for plots
    maxlabs <- merge(nabove, maxvals, by = "label")
    maxlabs$text <- paste("\U2191 N = ", maxlabs$measure.x, "\n Max = ", round(maxlabs$measure.y, digits = 1))
    maxlabs$measure <- highlimit
    maxlabs <- merge(maxlabs, unique(subdata[c("label", "Sex")]), by = "label")
  }
  # Generate plot for specific measure
  if (measure == "HbA1c") {
    plotcol <- cbPalette[2]
  } else {
    plotcol <- cbPalette[3]
  }
  if (all(subdata$N >= 10)) {
    x <- ggplot(subdata, aes(x = factor(label), y = measure)) +
      geom_violin(fill = plotcol) + geom_boxplot(width = 0.1) +
      facet_grid( ~ Sex, scales = "free") + theme_bw() + theme(legend.position = "none") + xlab("Genotype") +
      theme(axis.title = element_text(size = 30), axis.text = element_text(size = 30), legend.text = element_text(size = 30), strip.text = element_text(size = 30))
  } else {
    subdata2 <- subdata
    subdata2[subdata2$N < 10,]$measure <- 1000
    x <- ggplot(subdata[subdata$N >= 10,], aes(x = factor(label), y = measure)) +
      geom_violin(data = subdata2, fill = plotcol) + geom_boxplot(width = 0.1) +
      geom_dotplot(data = subdata[subdata$N < 10,], binaxis = "y", stackdir = "center", dotsize = 0.5, stackratio = 1.5) +
      facet_grid( ~ Sex, scales = "free") + theme_bw() + theme(legend.position = "none") + xlab("Genotype") +
      theme(axis.title = element_text(size = 30), axis.text = element_text(size = 30), legend.text = element_text(size = 30), strip.text = element_text(size = 30))
  }
  if (nrow(subdata[subdata$measure > highlimit,]) > 0) {
    x <- x + geom_text(data = maxlabs, aes(label = text), size = 8, hjust = -0.2, vjust = 1)
  }
  if (measure == "HbA1c") {
    x <- x + ylab("HbA1c (mmol/mol)") + scale_y_continuous(sec.axis = sec_axis(trans = ~./10.929 + 2.15, name = "HbA1c (%)"), limits = c(lowlimit, highlimit))
  } else {
    x <- x + ylab("Random glucose (mmol/L)") + ylim(lowlimit, highlimit)
  }
  return(x)
}

png("HbA1c_violin_plot_GnH.png", 1400, 850)
violinplot(measure = "HbA1c")
dev.off()

png("RG_violin_plot_GnH.png", 1400, 850)
violinplot(measure = "random_glucose")
dev.off()

## Generate summary characteristics for cohort

# Determine how many diagnosed with T2D since 2011
nrow(alldata[alldata$T2D == 1 & alldata$t2d_2011 == 1,])

## Apply exclusion criteria for age of T2D diagnosis analyses

# Make dataframe of those with T2D only (diagnosed at any time)
alldata <- alldata[alldata$T2D == 1 & !is.na(alldata$T2D_age),]

# Keep only those diagnosed with T2D since 2011
afterdata <- alldata[alldata$t2d_2011 == 1,]

### Compare age of T2D diagnosis ###

se <- function(x) sqrt(var(x, na.rm = TRUE)/length(x[!is.na(x)]))

# Compare age of T2D diagnosis across genotypes
tabdata <- data.frame(matrix(NA, nrow = 5, ncol = 9))
colnames(tabdata) <- c("Sex", "Genotype", "Status", "N", "Model.Beta", "Model.SE", "Model.LCI", "Model.UCI", "Model.P")
tabdata$Sex <- c(rep("Male", 2), rep("Female", 3))
tabdata$Genotype <- c(0, 1, 0:2)
tabdata$Status <- c("No mutations", "Hemizygote", "No mutations", "Heterozygote", "Homozygote")
for (i in 1:nrow(tabdata)) {
  subdata <- afterdata[afterdata$Sex == tabdata$Sex[i],]
  if (tabdata$Genotype[i] == 0) {
    if (length(unique(subdata$genotype)) > 1) {
      z <- lm(T2D_age ~ genotype + ethnicity, data = subdata)
      tabdata$Model.Beta[i] <- summary(z)$coefficients["genotype", "Estimate"]
      tabdata$Model.SE[i] <- summary(z)$coefficients["genotype", "Std. Error"]
      tabdata$Model.LCI[i] <- tabdata$Model.Beta[i] - (1.96*tabdata$Model.SE[i])
      tabdata$Model.UCI[i] <- tabdata$Model.Beta[i] + (1.96*tabdata$Model.SE[i])
      tabdata$Model.P[i] <- summary(z)$coefficients["genotype", "Pr(>|t|)"]
    }
  }
  subdata <- subdata[subdata$genotype == tabdata$Genotype[i],]
  tabdata$N[i] <- length(unique(subdata$NHS_ID))
}

write.table(tabdata, file = "sum_model_T2D_age_after2011_GnH.txt", col.names = TRUE, row.names = FALSE, sep = "\t", quote = FALSE)

# Generate model diagnostic plots for the regression models
mod_m_age_t2d <- lm(T2D_age ~ genotype + ethnicity, data = afterdata[afterdata$Sex == "Male",])
png("GnH_south_asian_male_diagnostic_plots.png", 800, 800)
par(mfrow = c(2, 2))
plot(mod_m_age_t2d, id.n = 0)
dev.off()

mod_f_age_t2d <- lm(T2D_age ~ genotype + ethnicity, data = afterdata[afterdata$Sex == "Female",])
png("GnH_south_asian_female_diagnostic_plots.png", 800, 800)
par(mfrow = c(2, 2))
plot(mod_f_age_t2d, id.n = 0)
dev.off()

